### PR TITLE
OADP-461: Document known issue: Restore of resources with Admission Webhooks

### PR DESCRIPTION
**Pull request by RichardHoch**
OADP 1.1.0, OCP 4.9+

Resolves: https://issues.redhat.com/browse/OADP-461 by adding a new section to the "Troubleshooting" section. 

Preview:  http://file.emea.redhat.com/rhoch/vel_admin_whook/backup_and_restore/application_backup_and_restore/troubleshooting.html#issues-with-velero-and-admission-workbooks

Please do not merge before Sept. 1.

### CODE REVIEW

File: modules/migration-debugging-velero-admin-webhooks.adoc
Comment (mperetzred) [inline modules/migration-debugging-velero-admin-webhooks.adoc:10]: maybe add link references as in the bug?
https://github.com/vmware-tanzu/velero/issues/4847

Comment (RichardHoch) [reply modules/migration-debugging-velero-admin-webhooks.adoc:10]: Added links for the Knative resources and IBM AppConnect resources issues.

Comment (sseago) [reply modules/migration-debugging-velero-admin-webhooks.adoc]: @RichardHoch The IBM issue was closed because there wasn't anything to fix in OADP. That's one of the issues that led to recognizing the need to document this sort of limitation.

File: modules/migration-debugging-velero-admin-webhooks.adoc
Comment (kaovilai) [inline modules/migration-debugging-velero-admin-webhooks.adoc:8]:
```suggestion
Velero has limited abilities to resolve admission webhook issues, so you might need to use an additional Velero plug-in or make changes to how you restore the workload.
```

**Review by mperetzred — APPROVED** [review]

**Comment by RichardHoch** [comment]
@kaovilai: Please review this PR again. i made the changes you suggested.

**Review by kaovilai — APPROVED** [review]

**Comment by RichardHoch** [comment]
@sbeskin-redhat: please review this PR.

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (sbeskin-redhat) [inline backup_and_restore/application_backup_and_restore/troubleshooting.adoc:30]: @RichardHoch 
Add 'Additional resources' after this line - see my comment below.

File: modules/migration-debugging-velero-admin-webhooks.adoc
Comment (sbeskin-redhat) [inline modules/migration-debugging-velero-admin-webhooks.adoc:14]: @RichardHoch 
"...requires you to res**t**ore the top level `Service` resource first..."

File: modules/migration-debugging-velero-admin-webhooks.adoc
Comment (sbeskin-redhat) [inline modules/migration-debugging-velero-admin-webhooks.adoc:24]: @RichardHoch 
We cannot have xrefs in modules. Make it into 'Additional resources' and a 'trailing include' in the Troubleshooting assembly following the include statement for this module. And in additional resources, we are now supposed to limit the text to the words in the square brackets. So this paragraph should appear in the Troubleshooting assembly like this:

include::modules/migration-debugging-velero-admin-webhooks.adoc[leveloffset=+1]

[role="_additional-resources"]
.Additional resources 
* xref:../../architecture/admission-plug-ins.adoc[Admission plug-ins]
* xref:../../architecture/admission-plug-ins#admission-webhooks-about_admission-plug-ins[Webhook admission plug-ins]
* xref:../../architecture/admission-plug-ins#admission-webhook-types_admission-plug-ins[Types of webhook admission plug-ins]

**Comment by RichardHoch** [comment]
@sbeskin-redhat : i made the changes you suggested. Please review this PR again.

File: modules/migration-debugging-velero-admin-webhooks.adoc
Comment (sbeskin-redhat) [reply modules/migration-debugging-velero-admin-webhooks.adoc:14]: @RichardHoch 
You've missed this one.

Comment (RichardHoch) [reply modules/migration-debugging-velero-admin-webhooks.adoc]: Fixed

**Comment by RichardHoch** [comment]
@sbeskin-redhat Fixed it...

**Comment by RichardHoch** [comment]
@sbeskin-redhat I removed the extra words. Now it should be OK.

**Comment by RichardHoch** [comment]
@sbeskin-redhat I removed the extra words. Now it should be OK.

**Review by mperetzred — APPROVED** [review]

**Comment by RichardHoch** [comment]
@sbeskin-redhat : I made the last changes you suggested -- removing extra words -- please review this PR again.

**Review by sbeskin-redhat — CHANGES_REQUESTED** [review]
@RichardHoch
I suspect that you didn't see my latest comments. They were marked 'pending'.

**Comment by RichardHoch** [comment]
@sbeskin-redhat -- The links work now. Anything else?

**Comment by sbeskin-redhat** [comment]
@RichardHoch 
Looks fine to me

File: modules/migration-debugging-velero-admin-webhooks.adoc
Comment (pneedle-rh) [inline modules/migration-debugging-velero-admin-webhooks.adoc:18]: You could update the replaceable value references so that they are lowercase and use underscores to separate the words. For example:

* `<restore_name>`
* `<backup_name>`

For more information, see [Code blocks, command syntax, and example output](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output) in the OpenShift documentation guidelines.

**Review by pneedle-rh — COMMENTED** [review]
@RichardHoch I have completed my review of this PR. Please let me know if you have any questions about my suggestions. Thanks!

File: modules/migration-debugging-velero-admin-webhooks.adoc
Comment (RichardHoch) [reply modules/migration-debugging-velero-admin-webhooks.adoc:5]: Done

**Comment by bergerhoffer** [comment]
The enterprise-4.12 label has been added to this PR.

This is because your PR targets the `main` branch and is labeled for enterprise-4.11. And any PR going into `main` must also target the latest version branch (enterprise-4.12).

If the update in your PR does NOT apply to version 4.12 onward, please re-target this PR to go directly into the appropriate version branch or branches (enterprise-4.x) instead of main.

File: modules/migration-debugging-velero-admission-webhooks.adoc
Comment (kalexand-rh) [inline modules/migration-debugging-velero-admission-webhooks.adoc:4]: I think that removing this blank line and the following line might help.

Also, this file needs to be in a different directory than `modules` - putting it elsewhere will likely fix the includes and let you place it in the `_topic_map.yml` file.

File: modules/migration-debugging-velero-admin-webhooks.adoc
Comment (RichardHoch) [reply modules/migration-debugging-velero-admin-webhooks.adoc]: @kaovilai @dymurray  Could you add some information here (or in the Jira) about how the creation of resources might help with the restore?

File: modules/migration-debugging-velero-admission-webhooks.adoc
Comment (RichardHoch) [reply modules/migration-debugging-velero-admission-webhooks.adoc:4]: Different directory was the answer (I removed the extra lines also).

File: _topic_maps/_topic_map.yml
Comment (kalexand-rh) [inline _topic_maps/_topic_map.yml:2349]: Your indentation is off, and it looks like you have some extra lines. This section should look like this:

```
- Name: Application backup and restore
  Dir: application_backup_and_restore
  Topics:
  - Name: OADP features and plugins
    File: oadp-features-plugins
  - Name: Issues with Velero and admission webhooks
    File: migration-debugging-velero-admission-webhooks
  - Name: Installing and configuring OADP
    Dir: installing
    Topics:
    - Name: About installing OADP
      File: about-installing-oadp
    - Name: About installing OADP
      File: about-installing-oadp
    - Name: Installing and configuring OADP with AWS
      File: installing-oadp-aws
    - Name: Installing and configuring OADP with Azure
      File: installing-oadp-azure
    - Name: Installing and configuring OADP with GCP
      File: installing-oadp-gcp
    - Name: Installing and configuring OADP with MCG
      File: installing-oadp-mcg
    - Name: Installing and configuring OADP with ODF
      File: installing-oadp-ocs
    - Name: Uninstalling OADP
      File: uninstalling-oadp
  - Name: Backing up and restoring
    Dir: backing_up_and_restoring
    Topics:
    - Name: Backing up applications
      File: backing-up-applications
    - Name: Restoring applications
      File: restoring-applications
  - Name: Troubleshooting
    File: troubleshooting
- Name: Control plane backup and restore
  Dir: control_plane_backup_and_restore
```

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (kalexand-rh) [inline backup_and_restore/application_backup_and_restore/troubleshooting.adoc:30]: You cannot include this file here. Nested assemblies are not allowed, include statements are not allowed in modules, and assemblies are not allowed in the modules directory.

If I remove this include statement and use the `_topic_map.yml` file that I described, the build passes.

File: modules/migration-debugging-velero-admission-webhooks.adoc
Comment (kalexand-rh) [inline modules/migration-debugging-velero-admission-webhooks.adoc:1]: You can't put an assembly in the `modules` directory. And modules cannot include other modules.

**Review by kalexand-rh — COMMENTED** [review]
I checked out a copy of this PR and was able to build locally after I made these changes.

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (RichardHoch) [inline backup_and_restore/application_backup_and_restore/troubleshooting.adoc:44]: @kaovilai  Can you give some information about how the creation of resources might help with the restore.

Comment (kaovilai) [reply backup_and_restore/application_backup_and_restore/troubleshooting.adoc:44]: If you restore/create the top level object first [like service.serving.knative.dev](https://github.com/knative/serving/issues/10926#issuecomment-886434177) then the child resources are typically created already and so it's more likely that admission webhook problems are resolved because child resources (typically blocked by these admission webhooks) do not need to be created by velero

**Comment by RichardHoch** [comment]
@pneedle-rh @kaovilai Please review this PR again.

File: _topic_maps/_topic_map.yml
Comment (RichardHoch) [inline _topic_maps/_topic_map.yml:2360]: @pneedle-rh I don't recall making this change. Should I remove it? Is it causing my build to break?

Comment (pneedle-rh) [reply _topic_maps/_topic_map.yml:2360]: @RichardHoch hi! In the `_topic_maps/_topic_map.yml` in your feature branch, these two lines are duplicated. This is causing the build failure.

Please remove the two duplicate lines and see if this resolves the issue.

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (sbeskin-redhat) [inline backup_and_restore/application_backup_and_restore/troubleshooting.adoc:44]: @RichardHoch 
top level -> top-level

**Comment by RichardHoch** [comment]
@pneedle-rh I made the changes you suggested just now. Please review this PR again.

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (pneedle-rh) [reply backup_and_restore/application_backup_and_restore/troubleshooting.adoc:47]: You could also potentially minimise the heading and the anchor ID by removing "Documented".

File: modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc
Comment (pneedle-rh) [inline modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc:19]: You could reverse the structure of this step to state the action first, then provide the reference to the Kubernetes documentation. For example:

```
. Examine the `.yaml` file of each `kind: MutatingWebhookConfiguration` to ensure that none of its rules block creation of the objects that are experiencing issues. For more information, see the link:https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#rulewithoperations-v1-admissionregistration-k8s-io[Kuberbetes documentation].
```

File: modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc
Comment (pneedle-rh) [inline modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc:23]: You could remove "still" from this line.

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (pneedle-rh) [inline backup_and_restore/application_backup_and_restore/troubleshooting.adoc:44]: I am not able to parse this line. Can you please clarify how this relates to the need to create a resource of a specific kind first?

**Review by pneedle-rh — COMMENTED** [review]
@RichardHoch I have now completed my additional review of this PR. Please let me know if you have any questions about my further suggestions.

Please also squash your commits into one.

Thanks!

File: modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc
Comment (RichardHoch) [reply modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc:8]: Sorry, my spell checker was disabled and i didn't catch these mistakes either.

File: backup_and_restore/application_backup_and_restore/migration-debugging-velero-admission-webhooks.adoc
Comment (RichardHoch) [reply backup_and_restore/application_backup_and_restore/migration-debugging-velero-admission-webhooks.adoc:1]: This was part of an experiment -- I wanted to nest assemblies, not realizing that is not allowed in openshift (it will be OK for Jupiter, btw). I deleted this file.

File: modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc
Comment (RichardHoch) [reply modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc:3]: No longer relevant.

File: modules/migration-debugging-velero-admission-webhooks-knative.adoc
Comment (RichardHoch) [reply modules/migration-debugging-velero-admission-webhooks-knative.adoc:3]: no longer relevant

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (RichardHoch) [reply backup_and_restore/application_backup_and_restore/troubleshooting.adoc:51]: Fixed.

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (kaovilai) [reply backup_and_restore/application_backup_and_restore/troubleshooting.adoc:38]: Creating a resource of a specific kind ie.. PizzaMaker automatically create child resources Pizza and DippingSauce which are sometimes has an admission webhook blocking creation without having a parent PizzaMaker created first.

**Comment by RichardHoch** [comment]
@pneedle-rh I've made the changes you've suggested. Please review this PR again.

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (RichardHoch) [reply backup_and_restore/application_backup_and_restore/troubleshooting.adoc]: @bergerhoffer Can I use the link that Tiger suggests?

Comment (GroceryBoyJr) [reply backup_and_restore/application_backup_and_restore/troubleshooting.adoc]: @RichardHoch I think your changes helped with what Paul was pointing out. Restoring the top-level object service.serving.knative.dev illustrates the problem.

**Review by GroceryBoyJr — COMMENTED** [review]
/LGTM

**Review by GroceryBoyJr — COMMENTED** [review]
LGTM

File: modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc
Comment (jboxman-rh) [inline modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc:16]: @RichardHoch, I think this is more correctly:

```
oc get mutatingwebhookconfigurations
```

As all lowercase. The uppercase works, but I think it's more of a convenience. The output is similar to:

```
NAME          WEBHOOKS   AGE
machine-api   2          215d
```

Comment (jboxman-rh) [inline modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc:16]:
```suggestion
$ oc get mutatingwebhookconfigurations
```

**Comment by RichardHoch** [comment]
@jboxman-rh Made the change -- please reply with L-G-T-M.

**Comment by jboxman-rh** [comment]
@RichardHoch, LGTM.

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (bergerhoffer) [inline backup_and_restore/application_backup_and_restore/troubleshooting.adoc:51]: This xref is missing an anchor

File: modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc
Comment (bergerhoffer) [inline modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc:19]: .yaml -> YAML (no markup)

```suggestion
. Examine the YAML file of each `kind: MutatingWebhookConfiguration` to ensure that none of its rules block creation of the objects that are experiencing issues. For more information, see link:https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#rulewithoperations-v1-admissionregistration-k8s-io[the official Kuberbetes documentation].
```

File: modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc
Comment (bergerhoffer) [inline modules/migration-debugging-velero-admission-webhooks-ibm-appconnect.adoc:21]:
```suggestion
. Check that any `spec.version` in `type: Configuration.appconnect.ibm.com/v1beta1` used at backup time is supported by the installed Operator.
```

File: modules/migration-debugging-velero-admission-webhooks-knative.adoc
Comment (bergerhoffer) [inline modules/migration-debugging-velero-admission-webhooks-knative.adoc:10]: Extra period

```suggestion
You can avoid such problems by restoring the top level `Service` resource first whenever you back up and restore Knative resources that use admission webhooks.
```

**Review by bergerhoffer — COMMENTED** [review]
Merge review for OADP GA: Several minor things that should be fixed at some point, but at the very least, the double period should be fixed before merging.

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (RichardHoch) [reply backup_and_restore/application_backup_and_restore/troubleshooting.adoc:51]: I want to point to the start of "Admissions plug-ins," not "About admission plug-ins"  -- that is, I want to reference the assembly "admission-plug-ins.adoc," not a module that is included in it. Do assemblies have anchors (or contexts)?

**Comment by RichardHoch** [comment]
@bergerhoffer I made all the changes you suggested except for adding an anchor to one of the xrefs -- please see my question there and review this PR. If I still need to an anchor, I'd like to do that in a separate later PR in the interest of time.

File: backup_and_restore/application_backup_and_restore/troubleshooting.adoc
Comment (bergerhoffer) [reply backup_and_restore/application_backup_and_restore/troubleshooting.adoc:51]: Hi @RichardHoch - yes assemblies still do have anchors, and all xrefs should have an anchor (even ones to assemblies). You'll have to check the actual assembly file to find the `id` at the top to use.

But I'm totally fine with merging this and following up on this in a separate PR.

**Review by bergerhoffer — COMMENTED** [review]
Merge review for OADP GA: LGTM now!

Can follow up on the xref anchor later in another PR.

**Pull request merged**
